plugins/animate/squeezimize.hpp

```diff
@@ -0,0 +1,388 @@
+                self->animation_geometry.width,
+                float((src_box.y - self->animation_geometry.y) + src_box.height) /
+                self->animation_geometry.height
+            };
+
+            const glm::vec4 target_box_pos{
+                float(self->minimize_target.x - self->animation_geometry.x) / self->animation_geometry.width,
+                float(self->minimize_target.y - self->animation_geometry.y) / self->animation_geometry.height,
+                float((self->minimize_target.x - self->animation_geometry.x) + self->minimize_target.width) /
+                self->animation_geometry.width,
+                float((self->minimize_target.y - self->animation_geometry.y) + self->minimize_target.height) /
+                self->animation_geometry.height
+            };
+
+            OpenGL::render_begin(target);
+            self->program.use(wf::TEXTURE_TYPE_RGBA);
+            self->program.uniformMatrix4f("matrix", target.get_orthographic_projection());
+            self->program.attrib_pointer("position", 2, 0, vertex_data_pos);
+            self->program.attrib_pointer("uv_in", 2, 0, vertex_data_uv);
+            self->program.uniform1i("upward", upward);
+            self->program.uniform1f("progress", progress);
+            self->program.uniform4f("src_box", src_box_pos);
+            self->program.uniform4f("target_box", target_box_pos);
+            self->program.set_active_texture(src_tex);
+            GL_CALL(glDrawArrays(GL_TRIANGLE_FAN, 0, 4));
+            OpenGL::render_end();
+        }
+    };
+
+    squeezimize_transformer(wayfire_view view,
+        wf::geometry_t minimize_target, wf::geometry_t bbox) : wf::scene::view_2d_transformer_t(view)
+    {
+        this->view = view;
+        this->minimize_target = minimize_target;
+        if (view->get_output())
+        {
+            output = view->get_output();
+            output->render->add_effect(&pre_hook, wf::OUTPUT_EFFECT_PRE);
+        }
+
+        animation_geometry.x     = std::min(bbox.x, minimize_target.x);
+        animation_geometry.y     = std::min(bbox.y, minimize_target.y);
+        animation_geometry.width =
+            std::max(std::max(std::max(bbox.width,
+                minimize_target.width),
+                (minimize_target.x + minimize_target.width) - bbox.x),
+                (bbox.x + bbox.width) - minimize_target.x);
+        animation_geometry.height =
+            std::max(std::max(std::max(bbox.height,
+                minimize_target.height),
+                (minimize_target.y + minimize_target.height) - bbox.y),
+                (bbox.y + bbox.height) - minimize_target.y);
+        OpenGL::render_begin();
+        program.compile(squeeze_vert_source, squeeze_frag_source);
+        OpenGL::render_end();
+    }
+
+    wf::geometry_t get_bounding_box() override
+    {
+        return this->animation_geometry;
+    }
+
+    wf::effect_hook_t pre_hook = [=] ()
```

Comment:
Is there a reason why this prehook is needed? The animation plugin is written in such a way that every animation gets this hook automatically set and managed by the main plugin: https://github.com/WayfireWM/wayfire/blob/0daf5c9a1f52383192d53bb20ad0d770581413f9/plugins/animate/animate.cpp#L86

In particular, this `animation_hook` automatically damages the view and calls the `step` method, so I think there should be no need at all for a pre-hook like this.

Reply:
It is needed because the animation geometry is larger than the view; it renders outside of the view's area. Thus, it must be damaged. You can see with the showrepaint plugin, and obviously, the animation renders outside of the view damage area.